tests/handlers/service_test.py

```diff
@@ -370,3 +375,93 @@ async def test_errors(client: AsyncClient) -> None:
+    assert r.status_code == 201
+
+    # Simple job list.
+    r = await client.get("/wobbly/jobs", headers=headers)
+    assert r.status_code == 200
+    expected = list(range(10))
+    expected.reverse()
+    assert [j["parameters"]["id"] for j in r.json()] == expected
+    assert "Link" not in r.headers
+
+    # Paginated queries.
+    r = await client.get("/wobbly/jobs", params={"limit": 5}, headers=headers)
+    assert r.status_code == 200
+    assert [j["parameters"]["id"] for j in r.json()] == expected[:5]
+    link_data = PaginationLinkData.from_header(r.headers["Link"])
+    assert not link_data.prev_url
+    assert link_data.first_url == "https://example.com/wobbly/jobs?limit=5"
+    assert link_data.next_url
+    r = await client.get(link_data.next_url, headers=headers)
+    assert r.status_code == 200
+    assert [j["parameters"]["id"] for j in r.json()] == expected[5:]
+    link_data = PaginationLinkData.from_header(r.headers["Link"])
+    assert not link_data.next_url
+    assert link_data.first_url == "https://example.com/wobbly/jobs?limit=5"
+    assert link_data.prev_url
+    r = await client.get(
+        link_data.prev_url, params={"limit": 1}, headers=headers
+    )
+    assert r.status_code == 200
+    assert [j["parameters"]["id"] for j in r.json()] == [expected[4]]
+
+    # Queries by phase.
+    r = await client.patch(
+        "/wobbly/jobs/1",
+        json={"phase": "QUEUED", "message_id": "some-message-id"},
+        headers=headers,
+    )
+    assert r.status_code == 200
+    r = await client.get("/wobbly/jobs", params={"limit": 5}, headers=headers)
+    assert r.status_code == 200
+    assert [j["parameters"]["id"] for j in r.json()] == expected[:5]
+    r = await client.get(
+        "/wobbly/jobs", params={"limit": 5, "phase": "QUEUED"}, headers=headers
+    )
+    assert r.status_code == 200
+    assert [j["parameters"]["id"] for j in r.json()] == [expected[9]]
+    link_data = PaginationLinkData.from_header(r.headers["Link"])
+    assert not link_data.next_url
+    assert not link_data.prev_url
+    r = await client.get(
+        "/wobbly/jobs", params={"phase": "PENDING"}, headers=headers
+    )
+    assert r.status_code == 200
+    assert [j["parameters"]["id"] for j in r.json()] == expected[:9]
+
+    # Queries with a creation time restriction.
+    stmt = select(SQLJob).where(SQLJob.id == 2)
+    async for session in db_session_dependency():
+        async with session.begin():
+            result = await session.execute(stmt)
+            job = result.scalar_one()
+            new_creation = datetime.now(tz=UTC) - timedelta(minutes=5)
+            job.creation_time = datetime_to_db(new_creation)
+    r = await client.get(
+        "/wobbly/jobs",
+        params={"since": (now - timedelta(seconds=1)).isoformat()},
+        headers=headers,
+    )
+    assert r.status_code == 200
+    since_expected = [*expected[:8], expected[9]]
+    assert [j["parameters"]["id"] for j in r.json()] == since_expected
```

Comment:
Some edge case tests (perhaps future work):

Edge cases:

- Query with limit=0 or limit=-1 (or even non-numeric) limit
- Pagination with limit larger than the total number of jobs
- Query without a phase parameter but with other unrelated parameters
- Creation time filtering: No jobs match the since parameter
- Test behaviour when no jobs exist

I think at first glance these would mostly be fine but perhaps good for regression testing

Reply:
Thanks, I've added several of these. I tend not to do full black box testing and instead do testing somewhat informed by the implementation so I don't test every combinatoric when the implementation is unrelated just to save from generating endless test cases, so I think for example the variations of phase are covered. I also tend to avoid testing basic Pydantic functionality (like non-numeric limit). But limit 0 and -1 are great edge cases, and I missed several of the pagination cases that you mentioned as well as several of the empty result tests. Those have now been added.